Seems like the imports are not sorted:

```
[ERROR] Failed to execute goal net.revelc.code:impsort-maven-plugin:1.2.0:check (check-imports) on project quarkus-spring-web: Imports are not sorted in /home/vsts/work/1/s/extensions/spring-web/runtime/src/main/java/io/quarkus/spring/web/runtime/ResponseContentTypeResolver.java -> [Help 1]
```

So what you need to do is basically do an `mvn process-resources` and the just commit the changes

--- a/extensions/spring-web/deployment/src/main/java/io/quarkus/spring/web/deployment/ControllerAdviceAbstractExceptionMapperGenerator.java
+++ b/extensions/spring-web/deployment/src/main/java/io/quarkus/spring/web/deployment/ControllerAdviceAbstractExceptionMapperGenerator.java
@@ -2,6 +2,7 @@ package io.quarkus.spring.web.deployment;
 
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Modifier;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -9,6 +10,7 @@ import java.util.Map;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
 
@@ -27,12 +29,19 @@ import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.spring.web.runtime.ResponseContentTypeResolver;
 import io.quarkus.spring.web.runtime.ResponseEntityConverter;
 
 class ControllerAdviceAbstractExceptionMapperGenerator extends AbstractExceptionMapperGenerator {
 
     private static final DotName RESPONSE_ENTITY = DotName.createSimple("org.springframework.http.ResponseEntity");
-    private static final DotName STRING = DotName.createSimple(String.class.getName());
+
+    // Preferred content types order for String or primitive type responses
+    private static final List<String> TEXT_MEDIA_TYPES = Arrays.asList(
+            MediaType.TEXT_PLAIN, MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML, MediaType.TEXT_XML);
+    // Preferred content types order for object type responses
+    private static final List<String> OBJECT_MEDIA_TYPES = Arrays.asList(
+            MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML, MediaType.TEXT_XML, MediaType.TEXT_PLAIN);
 
     private final MethodInfo controllerAdviceMethod;
     private final TypesUtil typesUtil;
@@ -41,6 +50,8 @@ class ControllerAdviceAbstractExceptionMapperGenerator extends AbstractException
     private final String declaringClassName;
 
     private final Map<Type, FieldDescriptor> parameterTypeToField = new HashMap<>();
+
+    private FieldDescriptor httpHeadersField;
 
     ControllerAdviceAbstractExceptionMapperGenerator(MethodInfo controllerAdviceMethod, DotName exceptionDotName,
             ClassOutput classOutput, TypesUtil typesUtil) {
@@ -104,89 +115,87 @@ class ControllerAdviceAbstractExceptionMapperGenerator extends AbstractException
                     "Parameter type " + parameterTypes.get(notAllowedParameterIndex).name() + " is not supported for method"
                             + controllerAdviceMethod.name() + " of class" + controllerAdviceMethod.declaringClass().name());
         }
+
+        createHttpHeadersField(cc);
+    }
+
+    private void createHttpHeadersField(ClassCreator classCreator) {
+        FieldCreator httpHeadersFieldCreator = classCreator
+                .getFieldCreator("httpHeaders", HttpHeaders.class)
+                .setModifiers(Modifier.PRIVATE);
+        httpHeadersFieldCreator.addAnnotation(Context.class);
+        httpHeadersField = httpHeadersFieldCreator.getFieldDescriptor();
     }
 
     @Override
     void generateMethodBody(MethodCreator toResponse) {
-        if (returnType.kind() == Type.Kind.VOID) {
-            AnnotationInstance responseStatusInstance = controllerAdviceMethod.annotation(RESPONSE_STATUS);
-
-            // invoke the @ExceptionHandler method
-            exceptionHandlerMethodResponse(toResponse);
-
-            // build a JAX-RS response
-            ResultHandle status = toResponse
-                    .load(responseStatusInstance != null ? getHttpStatusFromAnnotation(responseStatusInstance)
-                            : Response.Status.NO_CONTENT.getStatusCode());
-            ResultHandle responseBuilder = toResponse.invokeStaticMethod(
-                    MethodDescriptor.ofMethod(Response.class, "status", Response.ResponseBuilder.class, int.class),
-                    status);
-
-            ResultHandle httpResponseType = toResponse.load("text/plain");
-            toResponse.invokeVirtualMethod(
-                    MethodDescriptor.ofMethod(Response.ResponseBuilder.class, "type", Response.ResponseBuilder.class,
-                            String.class),
-                    responseBuilder, httpResponseType);
-
-            ResultHandle response = toResponse.invokeVirtualMethod(
-                    MethodDescriptor.ofMethod(Response.ResponseBuilder.class, "build", Response.class),
-                    responseBuilder);
-            toResponse.returnValue(response);
+        if (isVoidType(returnType)) {
+            generateVoidExceptionHandler(toResponse);
+        } else if (isEntityType(returnType)) {
+            generateResponseEntityExceptionHandler(toResponse);
         } else {
-            ResultHandle exceptionHandlerMethodResponse = exceptionHandlerMethodResponse(toResponse);
-
-            ResultHandle response;
-            if (RESPONSE_ENTITY.equals(returnType.name())) {
-                /*
-                 * By default we will send JSON back unless the ResponseEntity has a String body
-                 */
-                boolean addDefaultJsonContentType = true;
-                if (returnType.kind() == Type.Kind.PARAMETERIZED_TYPE) {
-                    if (returnType.asParameterizedType().arguments().size() == 1) {
-                        Type responseEntityParameterType = returnType.asParameterizedType().arguments().get(0);
-                        if (STRING.equals(responseEntityParameterType.name())) {
-                            addDefaultJsonContentType = false;
-                        }
-                    }
-                }
-
-                // convert Spring's ResponseEntity to JAX-RS Response
-                response = toResponse.invokeStaticMethod(
-                        MethodDescriptor.ofMethod(ResponseEntityConverter.class.getName(), "toResponse",
-                                Response.class.getName(), RESPONSE_ENTITY.toString(), boolean.class.getName()),
-                        exceptionHandlerMethodResponse, toResponse.load(addDefaultJsonContentType));
-            } else {
-                ResultHandle status = toResponse.load(getStatus(controllerAdviceMethod.annotation(RESPONSE_STATUS)));
-
-                ResultHandle responseBuilder = toResponse.invokeStaticMethod(
-                        MethodDescriptor.ofMethod(Response.class, "status", Response.ResponseBuilder.class, int.class),
-                        status);
-                /*
-                 * By default we will send JSON back unless the ResponseEntity has a String body
-                 */
-                if (!STRING.equals(returnType.name())) {
-                    toResponse.invokeVirtualMethod(
-                            MethodDescriptor.ofMethod(Response.ResponseBuilder.class, "type", Response.ResponseBuilder.class,
-                                    String.class),
-                            responseBuilder, toResponse.load(MediaType.APPLICATION_JSON));
-                }
-
-                toResponse.invokeVirtualMethod(
-                        MethodDescriptor.ofMethod(Response.ResponseBuilder.class, "entity", Response.ResponseBuilder.class,
-                                Object.class),
-                        responseBuilder, exceptionHandlerMethodResponse);
-
-                response = toResponse.invokeVirtualMethod(
-                        MethodDescriptor.ofMethod(Response.ResponseBuilder.class, "build", Response.class),
-                        responseBuilder);
-            }
-
-            toResponse.returnValue(response);
+            generateGenericResponseExceptionHandler(toResponse);
         }
     }
 
-    private ResultHandle exceptionHandlerMethodResponse(MethodCreator toResponse) {
-        String returnTypeClassName = returnType.kind() == Type.Kind.VOID ? void.class.getName() : returnType.name().toString();
+    private void generateVoidExceptionHandler(MethodCreator methodCreator) {
+        invokeExceptionHandlerMethod(methodCreator);
+        int status = getAnnotationStatusOrDefault(Response.Status.NO_CONTENT.getStatusCode());
+        ResultHandle result = new ResponseBuilder(methodCreator, status)
+                .withType(getResponseContentType(methodCreator, TEXT_MEDIA_TYPES))
+                .build();
+        methodCreator.returnValue(result);
+    }
+
+    private void generateResponseEntityExceptionHandler(MethodCreator methodCreator) {
+        ResultHandle result = methodCreator.invokeStaticMethod(
+                MethodDescriptor.ofMethod(ResponseEntityConverter.class.getName(), "toResponse",
+                        Response.class.getName(), RESPONSE_ENTITY.toString(), MediaType.class.getName()),
+                invokeExceptionHandlerMethod(methodCreator),
+                getResponseContentType(methodCreator, getSupportedMediaTypesForType(getResponseEntityType())));
+
+        methodCreator.returnValue(result);
+    }
+
+    private Type getResponseEntityType() {
+        if (isParameterizedType(returnType) && returnType.asParameterizedType().arguments().size() == 1) {
+            return returnType.asParameterizedType().arguments().get(0);
+        }
+        return returnType;
+    }
+
+    private void generateGenericResponseExceptionHandler(MethodCreator methodCreator) {
+        int status = getAnnotationStatusOrDefault(Response.Status.OK.getStatusCode());
+        ResultHandle result = new ResponseBuilder(methodCreator, status)
+                .withEntity(invokeExceptionHandlerMethod(methodCreator))
+                .withType(getResponseContentType(methodCreator, getSupportedMediaTypesForType(returnType)))
+                .build();
+
+        methodCreator.returnValue(result);
+    }
+
+    private List<String> getSupportedMediaTypesForType(Type type) {
+        if (isStringType(type) || isPrimitiveType(type)) {
+            return TEXT_MEDIA_TYPES;
+        }
+
+        return OBJECT_MEDIA_TYPES;
+    }
+
+    private ResultHandle getResponseContentType(MethodCreator methodCreator, List<String> supportedMediaTypeStrings) {
+        ResultHandle[] supportedMediaTypes = supportedMediaTypeStrings.stream()
+                .map(methodCreator::load)
+                .toArray(ResultHandle[]::new);
+
+        return methodCreator.invokeStaticMethod(
+                MethodDescriptor.ofMethod(ResponseContentTypeResolver.class, "resolve", MediaType.class,
+                        HttpHeaders.class, String[].class),
+                methodCreator.readInstanceField(httpHeadersField, methodCreator.getThis()),
+                methodCreator.marshalAsArray(String.class, supportedMediaTypes));
+    }
+
+    private ResultHandle invokeExceptionHandlerMethod(MethodCreator toResponse) {
+        String returnTypeClassName = isVoidType(returnType) ? void.class.getName() : returnType.name().toString();
 
         if (parameterTypes.isEmpty()) {
             return toResponse.invokeVirtualMethod(
@@ -228,10 +237,32 @@ class ControllerAdviceAbstractExceptionMapperGenerator extends AbstractException
         return toResponse.checkCast(bean, controllerAdviceMethod.declaringClass().name().toString());
     }
 
-    private int getStatus(AnnotationInstance instance) {
-        if (instance == null) {
-            return 200;
+    private int getAnnotationStatusOrDefault(int defaultValue) {
+        AnnotationInstance annotation = controllerAdviceMethod.annotation(RESPONSE_STATUS);
+        if (annotation == null) {
+            return defaultValue;
         }
-        return getHttpStatusFromAnnotation(instance);
+
+        return getHttpStatusFromAnnotation(annotation);
+    }
+
+    private boolean isVoidType(Type type) {
+        return Type.Kind.VOID.equals(type.kind());
+    }
+
+    private boolean isPrimitiveType(Type type) {
+        return Type.Kind.PRIMITIVE.equals(type.kind());
+    }
+
+    private boolean isStringType(Type type) {
+        return DotName.createSimple(String.class.getName()).equals(type.name());
+    }
+
+    private boolean isEntityType(Type type) {
+        return RESPONSE_ENTITY.equals(type.name());
+    }
+
+    private boolean isParameterizedType(Type type) {
+        return Type.Kind.PARAMETERIZED_TYPE.equals(type.kind());
     }
 }

--- a/extensions/spring-web/deployment/src/main/java/io/quarkus/spring/web/deployment/ResponseBuilder.java
+++ b/extensions/spring-web/deployment/src/main/java/io/quarkus/spring/web/deployment/ResponseBuilder.java
@@ -1,0 +1,46 @@
+package io.quarkus.spring.web.deployment;
+
+import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
+
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import io.quarkus.gizmo.MethodCreator;
+import io.quarkus.gizmo.ResultHandle;
+
+final class ResponseBuilder {
+
+    private final MethodCreator methodCreator;
+
+    private final ResultHandle delegate;
+
+    ResponseBuilder(MethodCreator methodCreator, int status) {
+        this.methodCreator = methodCreator;
+        this.delegate = withStatus(status);
+    }
+
+    public ResultHandle build() {
+        return methodCreator.invokeVirtualMethod(
+                ofMethod(Response.ResponseBuilder.class, "build", Response.class), delegate);
+    }
+
+    public ResponseBuilder withType(ResultHandle type) {
+        methodCreator.invokeVirtualMethod(
+                ofMethod(Response.ResponseBuilder.class, "type", Response.ResponseBuilder.class, MediaType.class),
+                delegate, type);
+        return this;
+    }
+
+    public ResponseBuilder withEntity(ResultHandle entity) {
+        methodCreator.invokeVirtualMethod(
+                ofMethod(Response.ResponseBuilder.class, "entity", Response.ResponseBuilder.class, Object.class),
+                delegate, entity);
+        return this;
+    }
+
+    private ResultHandle withStatus(int status) {
+        return methodCreator.invokeStaticMethod(
+                ofMethod(Response.class, "status", Response.ResponseBuilder.class, int.class),
+                methodCreator.load(status));
+    }
+}

--- a/extensions/spring-web/deployment/src/main/java/io/quarkus/spring/web/deployment/SpringWebProcessor.java
+++ b/extensions/spring-web/deployment/src/main/java/io/quarkus/spring/web/deployment/SpringWebProcessor.java
@@ -59,12 +59,6 @@ import io.quarkus.undertow.deployment.ServletInitParamBuildItem;
 
 public class SpringWebProcessor {
 
-    private static final DotName EXCEPTION = DotName.createSimple("java.lang.Exception");
-    private static final DotName RUNTIME_EXCEPTION = DotName.createSimple("java.lang.RuntimeException");
-
-    private static final DotName OBJECT = DotName.createSimple("java.lang.Object");
-    private static final DotName STRING = DotName.createSimple("java.lang.String");
-
     private static final DotName REST_CONTROLLER_ANNOTATION = DotName
             .createSimple("org.springframework.web.bind.annotation.RestController");
 
@@ -100,7 +94,7 @@ public class SpringWebProcessor {
     private static final DotName RESPONSE_ENTITY = DotName.createSimple("org.springframework.http.ResponseEntity");
 
     private static final Set<DotName> DISALLOWED_EXCEPTION_CONTROLLER_RETURN_TYPES = new HashSet<>(Arrays.asList(
-            MODEL_AND_VIEW, VIEW, MODEL, HTTP_ENTITY, STRING));
+            MODEL_AND_VIEW, VIEW, MODEL, HTTP_ENTITY));
 
     @BuildStep
     FeatureBuildItem registerFeature() {

--- a/extensions/spring-web/runtime/src/main/java/io/quarkus/spring/web/runtime/ResponseContentTypeResolver.java
+++ b/extensions/spring-web/runtime/src/main/java/io/quarkus/spring/web/runtime/ResponseContentTypeResolver.java
@@ -1,0 +1,52 @@
+package io.quarkus.spring.web.runtime;
+
+import static javax.ws.rs.core.HttpHeaders.ACCEPT;
+import static javax.ws.rs.core.MediaType.TEXT_PLAIN_TYPE;
+
+import java.util.List;
+import java.util.Objects;
+
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Variant;
+
+import org.jboss.resteasy.core.request.ServerDrivenNegotiation;
+
+public final class ResponseContentTypeResolver {
+
+    private static final MediaType DEFAULT_MEDIA_TYPE = TEXT_PLAIN_TYPE;
+
+    public static MediaType resolve(HttpHeaders httpHeaders, String... supportedMediaTypes) {
+        Objects.requireNonNull(httpHeaders, "HttpHeaders cannot be null");
+        Objects.requireNonNull(supportedMediaTypes, "Supported media types array cannot be null");
+
+        Variant bestVariant = getBestVariant(httpHeaders.getRequestHeader(ACCEPT), getMediaTypeVariants(supportedMediaTypes));
+
+        if (bestVariant != null) {
+            return bestVariant.getMediaType();
+        }
+
+        if (supportedMediaTypes.length > 0) {
+            return MediaType.valueOf(supportedMediaTypes[0]);
+        }
+
+        return DEFAULT_MEDIA_TYPE;
+    }
+
+    private static Variant getBestVariant(List<String> acceptHeaders, List<Variant> variants) {
+        ServerDrivenNegotiation negotiation = new ServerDrivenNegotiation();
+        negotiation.setAcceptHeaders(acceptHeaders);
+
+        return negotiation.getBestMatch(variants);
+    }
+
+    private static List<Variant> getMediaTypeVariants(String... mediaTypes) {
+        Variant.VariantListBuilder variantListBuilder = Variant.VariantListBuilder.newInstance();
+
+        for (String mediaType : mediaTypes) {
+            variantListBuilder.mediaTypes(MediaType.valueOf(mediaType));
+        }
+
+        return variantListBuilder.build();
+    }
+}

--- a/extensions/spring-web/runtime/src/main/java/io/quarkus/spring/web/runtime/ResponseEntityConverter.java
+++ b/extensions/spring-web/runtime/src/main/java/io/quarkus/spring/web/runtime/ResponseEntityConverter.java
@@ -17,26 +17,25 @@ import org.springframework.http.ResponseEntity;
  */
 public class ResponseEntityConverter {
 
-    public static Response toResponse(ResponseEntity responseEntity, boolean addJsonContentTypeIfNotSet) {
-        return new BuiltResponse(
-                responseEntity.getStatusCodeValue(), toHeaders(responseEntity.getHeaders(), addJsonContentTypeIfNotSet),
+    public static Response toResponse(ResponseEntity responseEntity, MediaType defaultContentType) {
+        return new BuiltResponse(responseEntity.getStatusCodeValue(),
+                addContentTypeIfMissing(toJaxRsHeaders(responseEntity.getHeaders()), defaultContentType),
                 responseEntity.getBody(),
                 new Annotation[0]);
     }
 
-    private static Headers<Object> toHeaders(HttpHeaders springHeaders, boolean addJsonContentTypeIfNotSet) {
+    private static Headers<Object> toJaxRsHeaders(HttpHeaders springHeaders) {
         Headers<Object> jaxRsHeaders = new Headers<>();
         for (Map.Entry<String, List<String>> entry : springHeaders.entrySet()) {
             jaxRsHeaders.addAll(entry.getKey(), entry.getValue().toArray(new Object[0]));
         }
-        /*
-         * We add the default application/json content type if no content type is specified
-         * since this is the default value when returning an object from a Spring RestController
-         */
-        if (addJsonContentTypeIfNotSet && !jaxRsHeaders.containsKey(javax.ws.rs.core.HttpHeaders.CONTENT_TYPE)) {
-            jaxRsHeaders.add(javax.ws.rs.core.HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
-        }
         return jaxRsHeaders;
     }
 
+    private static Headers<Object> addContentTypeIfMissing(Headers<Object> headers, MediaType contentType) {
+        if (!headers.containsKey(HttpHeaders.CONTENT_TYPE)) {
+            headers.add(HttpHeaders.CONTENT_TYPE, contentType);
+        }
+        return headers;
+    }
 }

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/CustomAdvice.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/CustomAdvice.java
@@ -1,7 +1,6 @@
 package io.quarkus.it.spring.web;
 
 import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
 
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
@@ -18,21 +17,35 @@ public class CustomAdvice {
 
     }
 
-    @ExceptionHandler(UnannotatedException.class)
+    @ExceptionHandler(HandledUnannotatedException.class)
     public void unannotatedException() {
 
     }
 
-    @ExceptionHandler(IllegalStateException.class)
-    public ResponseEntity<Error> handleIllegalStateException(IllegalStateException e,
-            HttpServletRequest request, HttpServletResponse response) {
-        return ResponseEntity.status(HttpStatus.PAYMENT_REQUIRED).header("custom-header", "custom-value")
-                .body(new Error(request.getRequestURI() + ":" + e.getMessage()));
+    @ExceptionHandler(HandledResponseEntityException.class)
+    public ResponseEntity<Error> handleResponseEntityException(HandledResponseEntityException e,
+            HttpServletRequest request) {
+
+        ResponseEntity.BodyBuilder bodyBuilder = ResponseEntity
+                .status(HttpStatus.PAYMENT_REQUIRED)
+                .header("custom-header", "custom-value");
+
+        if (e.getContentType() != null) {
+            bodyBuilder.contentType(e.getContentType());
+        }
+
+        return bodyBuilder.body(new Error(request.getRequestURI() + ":" + e.getMessage()));
     }
 
     @ResponseStatus(HttpStatus.EXPECTATION_FAILED)
-    @ExceptionHandler(IllegalArgumentException.class)
-    public Error handleGreetingIllegalArgumentException(IllegalArgumentException e) {
+    @ExceptionHandler(HandledPojoException.class)
+    public Error handlePojoExcepton(HandledPojoException e) {
         return new Error(e.getMessage());
+    }
+
+    @ResponseStatus(HttpStatus.I_AM_A_TEAPOT)
+    @ExceptionHandler(HandledStringException.class)
+    public String handleStringException(HandledStringException e) {
+        return e.getMessage();
     }
 }

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/Error.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/Error.java
@@ -1,8 +1,15 @@
 package io.quarkus.it.spring.web;
 
+import javax.xml.bind.annotation.XmlRootElement;
+
+@XmlRootElement
 public class Error {
 
-    private final String message;
+    private String message;
+
+    public Error() {
+
+    }
 
     public Error(String message) {
         this.message = message;
@@ -10,5 +17,9 @@ public class Error {
 
     public String getMessage() {
         return message;
+    }
+
+    public void setMessage(String message) {
+        this.message = message;
     }
 }

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/ExceptionThrowingController.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/ExceptionThrowingController.java
@@ -1,5 +1,6 @@
 package io.quarkus.it.spring.web;
 
+import org.springframework.http.MediaType;
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.RequestMapping;
@@ -10,49 +11,63 @@ import org.springframework.web.bind.annotation.RestController;
 @RequestMapping("/exception")
 public class ExceptionThrowingController {
 
-    @RequestMapping(method = RequestMethod.GET, path = "/first")
-    public void first() throws FirstException {
-        throw new FirstException("first");
+    @RequestMapping(method = RequestMethod.GET, path = "/unhandled/exception")
+    public void voidWithUnhandledAnnotatedException() throws UnhandledAnnotatedException {
+        throw new UnhandledAnnotatedException("unhandled");
     }
 
-    @GetMapping(path = "/second")
-    public String second() {
-        throw new SecondException();
+    @GetMapping(path = "/unhandled/runtime")
+    public String stringWithUnhandledAnnotatedRuntimeException() {
+        throw new UnhandledAnnotatedRuntimeException();
     }
 
-    @GetMapping("/void")
-    public void runtimeException() {
+    @GetMapping("/runtime")
+    public void voidWithRuntimeException() {
         throw new RuntimeException();
     }
 
     @GetMapping("/unannotated")
-    public void unannotated() {
-        throw new UnannotatedException();
+    public void voidWithUnannotatedException() {
+        throw new HandledUnannotatedException();
     }
 
-    @GetMapping("/responseEntity")
-    public Greeting handledByResponseEntity() {
-        throw new IllegalStateException("bad state");
+    @GetMapping("/re/re")
+    public ResponseEntity<Greeting> responseEntityWithResponseEntityException() {
+        throw new HandledResponseEntityException("bad state");
     }
 
-    @GetMapping("/responseEntityFromVoidReturningMethod")
-    public void handledByResponseEntityFromVoidReturningMethod() {
-        throw new IllegalStateException("bad state");
+    @GetMapping("/re/pojo")
+    public Greeting pojoWithResponseEntityException() {
+        throw new HandledResponseEntityException("bad state");
     }
 
-    @GetMapping("/pojo")
-    public Greeting greetingWithIllegalArgumentException() {
-        throw new IllegalArgumentException("hello from error");
+    @GetMapping("/re/void")
+    public void voidWithResponseEntityException() {
+        throw new HandledResponseEntityException("bad state");
     }
 
-    @GetMapping("/pojoWithVoidReturnType")
-    public void greetingWithIllegalArgumentExceptionAndVoidReturnType() {
-        throw new IllegalArgumentException("hello from error");
+    @GetMapping("/re/void/xml")
+    public void voidWithResponseEntityExceptionAsXml() {
+        throw new HandledResponseEntityException("bad state", MediaType.APPLICATION_XML);
     }
 
-    @GetMapping("/re")
-    public ResponseEntity<Greeting> responseEntityWithIllegalArgumentException() {
-        throw new IllegalStateException("hello from error");
+    @GetMapping("/pojo/re")
+    public ResponseEntity responseEntityWithPojoException() {
+        throw new HandledPojoException("bad state");
     }
 
+    @GetMapping("/pojo/pojo")
+    public Greeting pojoWithPojoException() {
+        throw new HandledPojoException("bad state");
+    }
+
+    @GetMapping("/pojo/void")
+    public void voidWithPojoException() {
+        throw new HandledPojoException("bad state");
+    }
+
+    @GetMapping("/string")
+    public String stringWithStringException() {
+        throw new HandledStringException("bad state");
+    }
 }

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/HandledPojoException.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/HandledPojoException.java
@@ -1,0 +1,8 @@
+package io.quarkus.it.spring.web;
+
+public class HandledPojoException extends RuntimeException {
+
+    public HandledPojoException(String message) {
+        super(message);
+    }
+}

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/HandledResponseEntityException.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/HandledResponseEntityException.java
@@ -1,0 +1,21 @@
+package io.quarkus.it.spring.web;
+
+import org.springframework.http.MediaType;
+
+public class HandledResponseEntityException extends RuntimeException {
+
+    private final MediaType contentType;
+
+    public HandledResponseEntityException(String message) {
+        this(message, null);
+    }
+
+    public HandledResponseEntityException(String message, MediaType contentType) {
+        super(message);
+        this.contentType = contentType;
+    }
+
+    public MediaType getContentType() {
+        return contentType;
+    }
+}

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/HandledStringException.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/HandledStringException.java
@@ -1,0 +1,8 @@
+package io.quarkus.it.spring.web;
+
+public class HandledStringException extends RuntimeException {
+
+    public HandledStringException(String message) {
+        super(message);
+    }
+}

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/HandledUnannotatedException.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/HandledUnannotatedException.java
@@ -1,0 +1,4 @@
+package io.quarkus.it.spring.web;
+
+public class HandledUnannotatedException extends RuntimeException {
+}

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/UnannotatedException.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/UnannotatedException.java
@@ -1,4 +1,0 @@
-package io.quarkus.it.spring.web;
-
-public class UnannotatedException extends RuntimeException {
-}

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/UnhandledAnnotatedException.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/UnhandledAnnotatedException.java
@@ -3,8 +3,8 @@ package io.quarkus.it.spring.web;
 import org.springframework.web.bind.annotation.ResponseStatus;
 
 @ResponseStatus
-public class FirstException extends Exception {
-    public FirstException(String message) {
+public class UnhandledAnnotatedException extends Exception {
+    public UnhandledAnnotatedException(String message) {
         super(message);
     }
 }

--- a/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/UnhandledAnnotatedRuntimeException.java
+++ b/integration-tests/spring-web/src/main/java/io/quarkus/it/spring/web/UnhandledAnnotatedRuntimeException.java
@@ -4,6 +4,6 @@ import org.springframework.http.HttpStatus;
 import org.springframework.web.bind.annotation.ResponseStatus;
 
 @ResponseStatus(code = HttpStatus.SERVICE_UNAVAILABLE)
-public class SecondException extends RuntimeException {
+public class UnhandledAnnotatedRuntimeException extends RuntimeException {
 
 }

--- a/integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/ExceptionHandlingIT.java
+++ b/integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/ExceptionHandlingIT.java
@@ -1,0 +1,7 @@
+package io.quarkus.it.spring.web;
+
+import io.quarkus.test.junit.NativeImageTest;
+
+@NativeImageTest
+public class ExceptionHandlingIT extends ExceptionHandlingTest {
+}

--- a/integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/ExceptionHandlingTest.java
+++ b/integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/ExceptionHandlingTest.java
@@ -1,0 +1,166 @@
+package io.quarkus.it.spring.web;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.emptyString;
+import static org.hamcrest.Matchers.is;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
+
+@QuarkusTest
+public class ExceptionHandlingTest {
+
+    @Test
+    public void testUnhandledAnnotatedException() {
+        RestAssured.when().get("/exception/unhandled/exception").then()
+                .contentType("text/plain")
+                .body(containsString("unhandled"))
+                .statusCode(500);
+    }
+
+    @Test
+    public void testUnhandledAnnotatedRuntimeException() {
+        RestAssured.when().get("/exception/unhandled/runtime").then()
+                .contentType("text/plain")
+                .body(is(emptyString()))
+                .statusCode(503);
+    }
+
+    @Test
+    public void testHandledRuntimeException() {
+        RestAssured.when().get("/exception/runtime").then()
+                .contentType("text/plain")
+                .body(is(emptyString()))
+                .statusCode(400);
+    }
+
+    @Test
+    public void testHandledRuntimeExceptionAsXml() {
+        RestAssured.given().accept("application/xml")
+                .when().get("/exception/runtime").then()
+                .contentType("application/xml")
+                .body(is(emptyString()))
+                .statusCode(400);
+    }
+
+    @Test
+    public void testHandledUnannotatedException() {
+        RestAssured.when().get("/exception/unannotated").then()
+                .contentType("text/plain")
+                .body(is(emptyString()))
+                .statusCode(204);
+    }
+
+    @Test
+    public void testHandledUnannotatedExceptionAsXml() {
+        RestAssured.given().accept("application/xml")
+                .when().get("/exception/unannotated").then()
+                .contentType("application/xml")
+                .body(is(emptyString()))
+                .statusCode(204);
+    }
+
+    @Test
+    public void testResponseEntityWithResponseEntityException() {
+        RestAssured.when().get("/exception/re/re").then()
+                .contentType("application/json")
+                .body(containsString("bad state"))
+                .body(containsString("/exception/re/re"))
+                .statusCode(402)
+                .header("custom-header", "custom-value");
+    }
+
+    @Test
+    public void testPojoWithResponseEntityException() {
+        RestAssured.when().get("/exception/re/pojo").then()
+                .contentType("application/json")
+                .body(containsString("bad state"))
+                .body(containsString("/exception/re/pojo"))
+                .statusCode(402)
+                .header("custom-header", "custom-value");
+    }
+
+    @Test
+    public void testVoidWithResponseEntityException() {
+        RestAssured.when().get("/exception/re/void").then()
+                .contentType("application/json")
+                .body(containsString("bad state"))
+                .body(containsString("/exception/re/void"))
+                .statusCode(402)
+                .header("custom-header", "custom-value");
+    }
+
+    @Test
+    public void testVoidWithResponseEntityExceptionAsXml() {
+        RestAssured.given().accept("application/xml")
+                .when().get("/exception/re/void").then()
+                .contentType("application/xml")
+                .body(containsString("bad state"))
+                .body(containsString("/exception/re/void"))
+                .statusCode(402)
+                .header("custom-header", "custom-value");
+    }
+
+    @Test
+    public void testVoidWithResponseEntityExceptionAsHardcodedXml() {
+        RestAssured.given().accept("application/json")
+                .when().get("/exception/re/void/xml").then()
+                .contentType("application/xml")
+                .body(containsString("bad state"))
+                .body(containsString("/exception/re/void/xml"))
+                .statusCode(402)
+                .header("custom-header", "custom-value");
+    }
+
+    @Test
+    public void testResponseEntityWithPojoException() {
+        RestAssured.when().get("/exception/pojo/re").then()
+                .contentType("application/json")
+                .body(containsString("bad state"))
+                .statusCode(417);
+    }
+
+    @Test
+    public void testPojoWithPojoException() {
+        RestAssured.when().get("/exception/pojo/pojo").then()
+                .contentType("application/json")
+                .body(containsString("bad state"))
+                .statusCode(417);
+    }
+
+    @Test
+    public void testVoidWithPojoException() {
+        RestAssured.when().get("/exception/pojo/void").then()
+                .contentType("application/json")
+                .body(containsString("bad state"))
+                .statusCode(417);
+    }
+
+    @Test
+    public void testVoidWithPojoExceptionAsXml() {
+        RestAssured.given().accept("application/xml")
+                .when().get("/exception/pojo/void").then()
+                .contentType("application/xml")
+                .body(containsString("bad state"))
+                .statusCode(417);
+    }
+
+    @Test
+    public void testStringWithStringException() {
+        RestAssured.when().get("/exception/string").then()
+                .statusCode(418)
+                .contentType("text/plain")
+                .body(is("bad state"));
+    }
+
+    @Test
+    public void testStringWithStringExceptionAsXml() {
+        RestAssured.given().accept("application/xml")
+                .when().get("/exception/string").then()
+                .statusCode(418)
+                .contentType("application/xml")
+                .body(is("bad state"));
+    }
+}

--- a/integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/SpringControllerTest.java
+++ b/integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/SpringControllerTest.java
@@ -1,8 +1,6 @@
 package io.quarkus.it.spring.web;
 
 import static org.hamcrest.Matchers.containsString;
-import static org.hamcrest.Matchers.emptyString;
-import static org.hamcrest.Matchers.is;
 
 import java.util.Optional;
 
@@ -126,83 +124,9 @@ public class SpringControllerTest {
     }
 
     @Test
-    public void testFirstResponseStatusHoldingException() {
-        RestAssured.when().get("/exception/first").then()
-                .contentType("text/plain")
-                .body(containsString("first"))
-                .statusCode(500);
-    }
-
-    @Test
-    public void testSecondResponseStatusHoldingException() {
-        RestAssured.when().get("/exception/second").then()
-                .contentType("text/plain")
-                .body(is(emptyString()))
-                .statusCode(503);
-    }
-
-    @Test
-    public void testExceptionHandlerVoidReturnType() {
-        RestAssured.when().get("/exception/void").then()
-                .contentType("text/plain")
-                .body(is(emptyString()))
-                .statusCode(400);
-    }
-
-    @Test
-    public void testExceptionHandlerWithoutResponseStatusOnExceptionOrMethod() {
-        RestAssured.when().get("/exception/unannotated").then()
-                .contentType("text/plain")
-                .body(is(emptyString()))
-                .statusCode(204);
-    }
-
-    @Test
-    public void testExceptionHandlerResponseEntityType() {
-        RestAssured.when().get("/exception/responseEntity").then()
-                .contentType("application/json")
-                .body(containsString("bad state"), containsString("responseEntity"))
-                .statusCode(402)
-                .header("custom-header", "custom-value");
-    }
-
-    @Test
-    public void testExceptionHandlerResponseEntityTypeFromVoidReturningMethod() {
-        RestAssured.when().get("/exception/responseEntityFromVoidReturningMethod").then()
-                .contentType("application/json")
-                .body(containsString("bad state"), containsString("responseEntity"))
-                .statusCode(402)
-                .header("custom-header", "custom-value");
-    }
-
-    @Test
-    public void testExceptionHandlerPojoEntityType() {
-        RestAssured.when().get("/exception/pojo").then()
-                .contentType("application/json")
-                .body(containsString("hello from error"))
-                .statusCode(417);
-    }
-
-    @Test
-    public void testControllerMethodWithVoidReturnTypeAndExceptionHandlerPojoEntityType() {
-        RestAssured.when().get("/exception/pojoWithVoidReturnType").then()
-                .contentType("application/json")
-                .body(containsString("hello from error"))
-                .statusCode(417);
-    }
-
-    @Test
     public void testRestControllerWithoutRequestMapping() {
         RestAssured.when().get("/hello").then()
                 .body(containsString("hello"));
-    }
-
-    @Test
-    public void testResponseEntityWithIllegalArgumentException() {
-        RestAssured.when().get("/exception/re").then()
-                .contentType("application/json")
-                .body(containsString("hello from error"))
-                .statusCode(402);
     }
 
     @Test

